python/lsst/daf/butler/registry/datasets/byDimensions/_manager.py

```diff
@@ -115,6 +115,22 @@ def refresh(self, *, universe: DimensionUniverse) -> None:
         self._byName = byName
         self._byId = byId
 
+    def deleteDatasetType(self, name: str) -> None:
+        # Docstring inherited from DatasetRecordStorageManager.
+        compositeName, componentName = DatasetType.splitDatasetTypeName(name)
+        if componentName is not None:
+            raise ValueError(f"Cannot delete a dataset type of a component of a composite (given {name})")
+        storage = self._byName.get(compositeName)
+        if storage is None:
+            raise KeyError(f"DatasetType {name} is not registered.")
+
+        # Delete the row
+        self._db.delete(self._static.dataset_type, ["name"], {"name": name})
```

Comment:
Should we catch IntegrityError here and raise something else? Should we even list a collection that uses this dataset type?

Reply:
Yeah, I would probably catch and raise just to at least say that this is because there is at least one dataset in that collection.  Actually querying for which dataset types for an even better error message is something we ought to do, but there's no easy way to do it now; maybe add a comment to [DM-24939](https://jira.lsstcorp.org/browse/DM-24939) to come back to this there?

Reply:
Oh, we should also make sure that catch _doesn't_ catch the case where the delete failed because the dataset type was already gone but this registry client had a stale cache (I don't think that'd be an integrity error, though).

Reply:
Deleting something that isn't there doesn't trigger an error -- it becomes a no-op so is safe. In that scenario though (where we have discovered that things are being deleted) it might be better to call the refresh anyhow.